feat(docs) add best development pratices (#12)
skore-io/culture: development-pratices.md

# Best development practices

> O documento a seguir descreve as regras de escrita nas linguagens de desenvolvimento no Skore que são utilizadas no projeto estruturado em Vuex. Nosso maior desafio é permitir que os desenvolvedores de front-end, façam que cada linha de código deva parecer ter sido escrita por uma única pessoa, não importa o número de contribuidores. Facilitando assim, a legibilidade e padrões que beneficiam o projeto como um todo, permitindo assim uma inclusão de novos contribuidores, a leitura e escrita no momento de desenvolvimento.

Reviewer: ```suggestion
O documento a seguir descreve as regras de escrita nas linguagens de desenvolvimento no Learning.Rocks, que são utilizadas no projeto estruturado em Nuxt/Vue. Nosso maior desafio é permitir que os desenvolvedores de front-end façam que cada linha de código deva parecer ter sido escrita por uma única pessoa, não importando o número de contribuidores. Isso tem como objetivo facilitar a legibilidade e manter padrões que beneficiam o projeto como um todo, permitindo assim uma fácil inclusão de novos contribuidores, e melhorando a leitura e a escrita no momento de desenvolvimento.
```

Este é um documento ativo e alterações podem ocorrer a qualquer momento.

## Sumário

1. [Commits](#commits)
1. [HTML](#html)
1. [Style](#style)
1. [JavaScript](#js)
1. [Traduções](#translations)

<a name="commits"></a>
## 1. Commits

> Para facilitar a contribuição de qualquer pessoa em um projeto, todas os commits, títulos de pull request devem estar em **inglês**.

Reviewer: ```suggestion
Para facilitar a contribuição de qualquer pessoa em um projeto, todas os commits e títulos de pull request devem estar em **inglês**.
```

A seguir existe uma ADR que detalha melhor como deve ser seguido esses padrões:
[Ler Padrões de Commits](https://github.com/skore-io/adrs/blob/master/doc/adr/0005-padroes-de-commits.md)

```javascript
// Good
git commit -m "feat(doc): add best development practices"

// Bad
git commit -m "Added Best development practices"
```

<a name="html"></a>
## 2. HTML

### Sumário HTML

1. [Sintaxe HTML](#html-syntax)
1. [Comentários HTML](#html-comments)

<a name="html-syntax"></a>
### 2.1. Sintaxe HTML

Matenha o formato do código com dois espaços. Você pode configurar seu editor para isso.

```html
<!-- Good -->
<Row>
  <UserList>
    <li class="nav-item">
      <a class="nav-link">

<!-- Bad-->
<Row>
      <UserList>
            <li class="nav-item">
                  <a class="nav-link">
```

Sempre use aspas duplas.

```html
<!-- Good -->
<Button class="primary" />

<!-- Bad-->
<Button class='primary' />
```

> Separe o elemento do bloco por uma linha em branco e agrupe os elementos internos do bloco.

Reviewer: Cláusula anti @ighorfmelo

Author: HAHAHAHA

Reviewer 2: Olha os caras kkkkkkkk

```html
<style lang="scss" scoped>
 ...
</style>
```

Matenha o formato do código com dois espaços. Você pode configurar seu editor para isso.

```css
/* Good */
.nav-item {
  display: inline-block;
  margin: 0 5px;
}

/* Bad */
.nav-item {
    display: inline-block;
    margin: 0 5px;
}
```

Sempre use aspas duplas para manter a semântica.

```css
/* Good */
[type="text"]
[class^="..."]

.nav-item:after {
  content: "bilú";
}

/* Bad */
[type='text']
[class^='...']

.nav-item:after {
  content: 'bilú';
}
```

Inclua um único espaço antes do colchete de abertura de um conjunto de regras.

```css
/* Good */
.header {
  ...
}

/* Bad */
.header{
  ...
}
```

Inclua um único espaço após os dois pontos de uma declaração.

```css
/* Good */
.header {
  margin-bottom: 20px;
}

/* Bad */
.header{
  margin-bottom:20px;
}
```

Inclua um ponto e vírgula (;) no final da última declaração em seu bloco.

```css
/* Good */
.header {
  margin-bottom: 20px;
}

/* Bad */
.header{
  margin-bottom:20px
}
```

Mantenha uma declaração por linha.

```css
/* Good */
.selector-1,
.selector-2,
.selector-3 {
  ...
}

/* Bad */
.selector-1, .selector-2, .selector-3 {
  ...
}
```

> Separe cada conjunto de regras por uma linha em branco, assim como seu aninhamento.

Reviewer: ```suggestion
Separe cada conjunto de regras por uma linha em branco, assim como seu alinhamento.
```

  ...
}
```

<a name="css-order"></a>
### 3.2. Ordem de declaração

O ideal é que as declarações de propriedades sejam adicionadas em ordem alfabética, facilitando a leitura.

```css
/* Good */
.selector-1 {
  background: #fff;
  border: 1px solid #333333;
  color: #333;
  display: block;
  height: 200px;
  margin: 5px;
  padding: 5px;
  width: 200px;
}

/* Bad */
.selector-1 {
  padding: 5px;
  width: 200px;
  height: 200px;
  background: #fff;
  margin: 5px;
  color: #333;
  border: 1px solid #333333;
  display: block;
}
```

<a name="css-class-name"></a>
### 3.3. Nome de classe CSS

Mantenha a classe em letras minúsculas e use hífens para separar o nome da classe.

```css
/* Good */
.page-header { ... }

/* Bad */
.pageHeader { ... }
.page_header { ... }
```

Evite dar nomes muito curtos para a classe e sempre escolha nomes significativos que forneçam a função da classe.

```css
/* Good */
.btn { ... }
.page-header { ... }
.progress-bar { ... }

/* Bad */
.s { ... }
.ph { ... }
.block { ... }
```

<a name="css-performance"></a>
### 3.4. Performance CSS

Evite usar IDs.

```css
/* Good */
.header { ... }
.section { ... }

/* Bad */
#header { ... }
#section { ... }
```

> Evite aninhamento com mais de três elementos, isso deixa mais complexo o seu código, de preferência sempre use classes únicas nestes casos.

Reviewer: ```suggestion
Evite alinhamento com mais de três elementos, isso deixa mais complexo o seu código, de preferência sempre use classes únicas nestes casos.
```

Author: Seria aninhamento (nesting), a ideia.

Reviewer: Eita, nem conhecia essa palavra hahah. Acho melhor por _nesting_ em inglês mesmo 😄

Author: Hahahaha, acho melhor mesmo... Fica melhor entendido pra todos.

```css
/* Good */
.navbar-link { ... }

/* Bad */
.navbar {
  ul {
    li {

>       a { ... }

Reviewer: Isso é um pouco discutível já que usamos bastante o nesting do `scss`, mas podemos conversar melhor sobre. Acho muito ruim visualizar e revisar o código quando há vários graus de nesting. Inclusive há um warning sobre isso no próprio site deles hahah: 

![image](https://user-images.githubusercontent.com/65301553/207667306-c0d3a5aa-c546-4742-af3d-aa32a5da6355.png)

Author: Exatamate... O que observei que tem bastante hoje no projeto, é não utilizar classes que vão direto no elemento, em vez disso existe muitos graus de nesting nos códigos, chega a ficar até difícil de debugar. Existe outros pontos negativos que é possível dizer que é gerado em relação a isso, um deles é esse print que fala da renderização no browser. Então sempre que houver a necessidade de estilizar algo que tem um nó mais fundo, é melhor optar pra usar uma classe pra ele.

Reviewer: Boa. É que a gente gosta de seguir no `scss` a estrutura do HTML né. Então se há 4 elementos, cada um filho do anterior, a gente acaba fazendo nesting desses 4 elementos no `scss`. Mas se, por exemplo, a gente só precisar adicionar uma classe ao último filho, poderíamos já chamá-lo direto nos níveis acima...

Author: Isso, acho o mais certo a ser feito!

Reviewer 3: Boa demais, fomentando apenass um pouco mais a discussão, acho que na maioria das vezes quando temos muito nesting é um bom sinal de que o componente está com um conhecimento bem grande. Não é regra, mas é algo que podemos ficar de 👀 nos PR's.

```css
/* Good */
.navbar {
  margin-bottom: 20px;

  @media (min-width: 992px) {
    left: 0;
    position: fixed;
    top: 0;
  }
}

/* Bad */
.navbar {
  margin-bottom: 20px;
  left: 0;
  position: fixed;
  top: 0;

  @media (max-width: 991px) {
    position: static;
  }
}
```

<a name="js"></a>
## 4. Scripts

Atualmente são utilizadas em nosso projeto bibliotecas como ESLint e Prettier, então muitas regras de códigos são corrigidas automaticamente ao utiliza-las no editor de código. Nos assuntos abaixo, são regras personalizadas que devem ser observadas no momento do desenvolvimento, sendo elas seguidas pela equipe.

### Sumário

1. [4.1. Sintaxe do script](#js-syntax)
1. [4.2. Variáveis](#js-variables)
1. [4.3. Comentários](#js-comments)
1. [4.4. Imports](#js-comments)

<a name="js-syntax"></a>
### 4.1. Sintaxe do script

Matenha o formato do código com dois espaços. Você pode configurar seu editor para isso.

```js
// Good
try {
  code...
} catch (e) {
  code...
}

// Bad
try {
    code...
} catch (e) {
    code...
}
```

Para várias condicionais, sempre use colchetes {}.

```js
// Good
if (condition) {
  statement

> } else if (condition) {

Reviewer: Sim. Acho que isso segue o paradigma de 'programação funcional', de cada função ter uma única responsabilidade.

  statement
} else {
  statement
}

// Bad
if (condition) statement
else if (condition) statement

> else statement

Reviewer: Acabaram com o @ighorfmelo de novo hahahah

Reviewer 2: Por mim eu usava colchetes em tudo, vocês que gostam de codar sem.

```

Para verificações de igualdade estrita, === deve ser usado em favor de ==.

```js
// Good
if (foo === 'foo') {
  statement
}

// Bad
if (foo == 'foo') {
  statement
}
```

<a name="js-variables"></a>
### 4.2. Variáveis

Todas as variáveis devem ser declaradas antes de serem usadas.

```js
// Good
const { name, age } = this.getUser(this.id)
const minAge = 18

> console.log(`Oi ${name}, bem-vindo.`)

Reviewer: O `minAge` não deveria ser usado dentro desse `console` para exemplificar melhor? Porque caso ele fosse usado em outro lugar lá embaixo, às vezes é melhor declará-lo lá, próximo de sua utilização.

Reviewer: Sim. Concordo. Depende de onde ele será utilizado dentro da função!

...

// Bad
const { name, age } = this.getUser(this.id)

console.log(`Oi ${name}, bem-vindo.`)

const minAge = 18
...

```

<a name="js-comments"></a>

> ### 4.3. Comentários

Reviewer: Um padrão legal de comentários é o da extensão **Better Comments**: https://marketplace.visualstudio.com/items?itemName=aaron-bond.better-comments

![image](https://user-images.githubusercontent.com/65301553/207668377-8b967c44-6350-4125-b27c-2f0bf0dd6ac1.png)

Author: Boa! Acho que vale a pena verificar a extensão pra todos utilizarem da mesma forma.

Use apenas comentários caso o código terá uma futura atualização em sua estrutura, onde será removido posteriomente.

```js
try {
  // TODO: Add mutation to update a class
} catch (e) {
  this.$toast.error(this.$t('error'))
}
```

### 4.4. Imports

Sempre utiliza caminhos absolutos na chamada de outros arquivos, evitando refatorações caso o arquivo seja movido de pasta e facilitando entender de onde está sendo sua origem.

```js
/* Good */
import Modal from '@/components/UI/Modal'
import TextField from '@/components/UI/TextField'
import AudienceModal from '@/modules/players/LiveCourse/components/LiveClass/Modal/Audience/Enrollment/Type'

/* Bad */
import Modal from '../../../../../../../../components/UI/Modal'
import TextField from '@/components/UI/TextField'
import AudienceModal from '../Enrollment/Type'

```

<a name="translations"></a>
## 5. Traduções

> Sempre utilize o padrão de abertura de tags para i18n com a linguagem em yaml. A sequência de traduções atual é en, pt e es.

Reviewer: Podemos debater se não é melhor começar com `pt`. Há uma galera começando por ele...

Author: Sim... Eu só deixei assim porque maioria dos componentes que vejo hoje começa com `en`. Então talvez vale estudar se é melhor começar com nossa lingua materna. Porém, os componentes que estão no padrão do exemplo, precisariam ser todos refatorados pra começar em português, ou não mexeriamos? Iria gerar um PR só pra isso.

Reviewer: Sim, se fossemos padronizar com `pt` primeiro, teríamos que subir um PR gigante arrumando tudo. Ou podemos só adotar a partir de agora e futuramente fazemos isso...

> Sempre utilize o padrão de abertura de tags para i18n com a linguagem em yaml. A sequência de traduções atual é en, pt e es.

Reviewer 2: Podemos adicionar como regra também a divisão de escopo na tradução. 
Por exemplo, evitar o uso de dropdown_label, dropdown_description e usar
```
dropdown:
  label:
  description:
```

Author: Ótima sugestão! Vou adicionar isso.

```html
<i18n lang="yaml">
en:
  title: Manage the list
pt:
  title: Gerencie a lista
es:
  title: Gestione la lista
</i18n>
```

### 5.1. Aspas

> Evite utiliza-las em todas as traduções, a menos que a mesma possui outras aspas dentro da frase.

Reviewer: Sinceramente eu prefiro usar aspas em todas as traduções. Acho que fica mais padronizado do que ficar usando só quando há outras aspas dentro da frase. Fora que as aspas dão uma impressão de `string`. O que vocês acham?

Reviewer 2: Eu também prefiro as aspas.

Author: Sim! Acho que vai do gosto de cada um como é utilizado atualmente. Isso foi uma discursão que tive inclusive com o @guiwb em nossa 1:1. Pois ele por exemplo exemplificou que prefere sem aspas... Então acho que vale a pena todos entrarem em um acordo de como materiamos pra utilizar.

Reviewer 3: Sou mesmo e se visse isso a tempo adivinhem, eu iria contra a regra de colocar aspas, mas vejo um senso comum entre vocês e não vou me opor à essa decisão hahaha
